File: tools/binding_compliance/conformance/families/registry_accessors.py
"""Operation-scoped registry accessors over disposable process-owned state."""

import json
from collections.abc import Mapping
from functools import partial
from pathlib import Path
from typing import Any

from ..coverage import CoveragePredicate, FamilyCoveragePolicy

OPERATIONS = {
    "registry-game": {
        "set_game": ("set_game", "registrySetGame", "registry_set_game"),
        "get_game": ("get_game", "registryGetGame", "registry_get_game"),
        "GAME": ("registry_key_game",),
    },
    "registry-gui": {
        "is_gui_mode": ("is_gui_mode", "registry_is_gui_mode"),
        "IS_GUI_MODE": ("registry_key_is_gui_mode",),
    },
    "registry-paths": {
        "set_application_dir": ("set_application_dir", "setApplicationDir"),
        "get_application_dir": ("get_application_dir", "getApplicationDir"),
    },
    "registry-context": {
        name: (name,)
        for name in (
            "get_yaml_cache",
            "get_manual_docs_gui",
            "get_game_path_gui",
            "get_local_dir",
            "is_version_auto_detected",
            "is_xse_valid",
            "is_enb_present",
            "get_game_version_string",
        )
    },
}


def _observed(family: str, value: Mapping[str, Any]) -> bool:
    """Require each public result and reset effect without accepting extra fields."""
    if family == "registry-game":
        return (
            set(value) == {"key", "game", "replacement", "afterClearPresent"}
            and value["key"] == "gamevars_game"
            and all(isinstance(value[k], str) for k in ("game", "replacement"))
            and value["afterClearPresent"] is False
        )
    if family == "registry-gui":
        return (
            set(value) == {"key", "default", "enabled", "disabled", "afterClear"}
            and value["key"] == "is_gui_mode"
            and (
                value["default"] is False
                and value["enabled"] is True
                and value["disabled"] is False
                and value["afterClear"] is False
            )
        )
    if family == "registry-paths":
        return set(value) == {
            "initial",
            "path",
            "replacement",
            "afterClear",
            "files",
        } and (
            value["initial"] is None
            and value["afterClear"] is None
            and value["files"] == []
            and all(isinstance(value[k], str) for k in ("path", "replacement"))
        )
    return set(value) == {"initial", "stored", "afterClear", "localDir", "files"} and (
        value["files"] == []
        and isinstance(value["localDir"], str)
        and all(_context(value[k]) for k in ("initial", "stored", "afterClear"))
    )


def _context(value: Any) -> bool:
    """Preserve nullable object values and typed flags separately from absence."""
    return (
        isinstance(value, dict)
        and set(value)
        == {
            "yamlCache",
            "manualDocs",
            "gamePath",
            "autoDetected",
            "xseValid",
            "enbPresent",
            "version",
        }
        and all(
            value[k] is None or isinstance(value[k], str)
            for k in (
                "yamlCache",
                "manualDocs",
                "gamePath",
            )
        )
        and all(
            type(value[k]) is bool
            for k in (
                "autoDetected",
                "xseValid",
                "enbPresent",
            )
        )
        and isinstance(value["version"], str)
    )


def validate_pack(document: Mapping[str, Any], root: Path) -> tuple[Path, ...]:
    """Reject undeclared input and paths before any adapter touches registry state."""
    family = document["familyId"]
    if family not in OPERATIONS:
        raise ValueError("unknown registry accessor family")
    fixture_root = (root / document["fixtureRoot"]).resolve()
    paths = []
    for case in document["scenarios"]:
        reference = case["input"].get("fixtureRef")
        if (
            case["action"] != family + ".observe"
            or case["input"] != {"fixtureRef": reference}
            or case["fixtureRefs"] != [reference]
        ):
            raise ValueError("registry accessor needs its sole declared input fixture")
        path = (fixture_root / document["fixtures"][reference]).resolve()
        if not path.is_relative_to(fixture_root):
            raise ValueError("registry fixture escapes root")
        fixture = json.loads(path.read_text(encoding="utf-8"))
        if not isinstance(fixture, dict) or set(fixture) != {"request"}:
            raise ValueError("registry fixture must be input-only")
        request = fixture["request"]
        keys = (
            {"game", "replacement"}
            if family == "registry-game"
            else {"value", "version"}
            if family == "registry-context"
            else set()
        )
        if (
            not isinstance(request, dict)
            or set(request) != keys
            or not all(isinstance(v, str) for v in request.values())
            or not _observed(family, case["expected"])
        ):
            raise ValueError("malformed registry accessor request or observation")
        paths.append(path)
    return tuple(paths)


def coverage_policy(family: str) -> FamilyCoveragePolicy:
    """Credit exact executed convenience methods, never their entire owner module."""
    return FamilyCoveragePolicy(
        family,
        tuple(
            CoveragePredicate(
                id=family
                + "."
                + ("key-" if symbol.isupper() else "")
                + symbol.replace("_", "-").lower(),
                capability_id=family + ".observe",
                action=family + ".observe",
                observation_family="registry-state",
                rust_symbols=(symbol,),
                runtime_operations=(None, *aliases),
                matches=partial(_observed, family),
            )
            for symbol, aliases in OPERATIONS[family].items()
        ),
    )

File: business-logic/classic-scanlog-core/tests/semantic_conformance/registry_accessors.rs
//! Registry convenience calls in an isolated receipt process.

use super::{RunnerResult, invalid, text};
use classic_registry_core::{self as registry, Keys};
use serde_json::{Value, json};
use std::path::Path;

/// Clear process-global state even when path conversion or observation fails.
struct Reset;
impl Drop for Reset {
    fn drop(&mut self) {
        registry::clear_all();
    }
}

/// Read nullable references and default-valued preferences through public getters.
fn context() -> Value {
    json!({"yamlCache": registry::get_yaml_cache::<String>(),
        "manualDocs": registry::get_manual_docs_gui::<String>(),
        "gamePath": registry::get_game_path_gui::<String>(),
        "autoDetected": registry::is_version_auto_detected(),
        "xseValid": registry::is_xse_valid(), "enbPresent": registry::is_enb_present(),
        "version": registry::get_game_version_string()})
}

/// Exercise one owner contract; all filesystem paths belong to a temporary directory.
pub(super) fn execute(family: &str, fixture: &Value) -> RunnerResult<Value> {
    registry::clear_all();
    let _reset = Reset;
    let request = &fixture["request"];
    if family == "registry-game" {
        registry::set_game(text(&request["game"])?);
        let game = registry::get_game();
        registry::set_game(text(&request["replacement"])?);
        let replacement = registry::get_game();
        registry::clear_all();
        return Ok(
            json!({"key":Keys::GAME,"game":game, "replacement":replacement,
            "afterClearPresent":registry::is_registered(Keys::GAME)}),
        );
    }
    if family == "registry-gui" {
        let default = registry::is_gui_mode();
        registry::register(Keys::IS_GUI_MODE, true);
        let enabled = registry::is_gui_mode();
        registry::register(Keys::IS_GUI_MODE, false);
        let disabled = registry::is_gui_mode();
        registry::clear_all();
        return Ok(
            json!({"key":Keys::IS_GUI_MODE,"default":default,"enabled":enabled,"disabled":disabled,
            "afterClear":registry::is_gui_mode()}),
        );
    }
    let workspace = tempfile::tempdir()?;
    let root = workspace.path();
    if family == "registry-paths" {
        let initial = registry::get_application_dir();
        registry::set_application_dir(root.join("first"));
        let path = relative(
            root,
            &registry::get_application_dir().ok_or_else(|| invalid("missing override"))?,
        )?;
        registry::set_application_dir(root.join("second"));
        let replacement = relative(
            root,
            &registry::get_application_dir().ok_or_else(|| invalid("missing override"))?,
        )?;
        registry::clear_all();
        return Ok(
            json!({"initial":initial,"path":path,"replacement":replacement,
            "afterClear":registry::get_application_dir(),"files":files(root)?}),
        );
    }
    if family != "registry-context" {
        return Err(invalid("unknown registry accessor family").into());
    }
    let initial = context();
    for key in [Keys::YAML_CACHE, Keys::MANUAL_DOCS_GUI, Keys::GAME_PATH_GUI] {
        registry::register(key, text(&request["value"])?);
    }
    // Python treats empty version strings as absent; keep reference emptiness independent.
    registry::register(Keys::GAME_VERSION, text(&request["version"])?);
    for key in [
        Keys::VERSION_AUTO_DETECTED,
        Keys::XSE_VALID,
        Keys::ENB_PRESENT,
    ] {
        registry::register(key, true);
    }
    registry::register(Keys::LOCAL_DIR, root.join("local"));
    let local = relative(root, &registry::get_local_dir())?;
    let stored = context();
    registry::clear_all();
    Ok(
        json!({"initial":initial,"stored":stored,"afterClear":context(),
        "localDir":local,"files":files(root)?}),
    )
}

/// Reject an escaped public path instead of normalizing away a wrong directory.
fn relative(root: &Path, path: &Path) -> RunnerResult<String> {
    Ok(path
        .strip_prefix(root)?
        .to_string_lossy()
        .replace('\\', "/"))
}

/// Retain any forbidden directory entry instead of assuming setters never write.
fn files(root: &Path) -> RunnerResult<Vec<String>> {
    let mut entries = std::fs::read_dir(root)?
        .map(|entry| Ok(entry?.file_name().to_string_lossy().into_owned()))
        .collect::<std::io::Result<Vec<_>>>()?;
    entries.sort();
    Ok(entries)
}
